File: quickSort.c
#include <stdio.h>
#include "quickSort.h"
int *list;
int n;

void pivote(int, int , int);
void sort(int, int);
void quicksort(int x[20], int y) {
    list = x;
    n = y;

    sort(0,n-1);
} 

void pivote(int p, int q, int r) {
    sort(p, q-1);
    sort(q+1, r);
}

void sort(int p, int r) {
    if(p < r) {
        int i,j,q,temp,x;

        x = list[r];
        q = p;
        for(i=p; i<=r; i++) {
            if(list[i] <= x) {
                temp = list[i];
                list[i] = list[q];
                list[q] = temp;
                q++;
            }
        }

        pivote(p,q-1,r);
    }
}

File: quickSort.java
public class quickSort extends sortArray{
    static void swap(int[] ar, int i, int j) {
        int t;
        t = ar[i];
        ar[i] = ar[j];
        ar[j] = t;
    }
    
    public void sort(int[] list) {
        sortList(list, 0, list.length-1);
    }

    private static void partition(int[] list, int p, int q, int r) {
        sortList(list, p, q-1);
        sortList(list, q+1, r);
    }

    private static void sortList(int[] list, int p, int r) {
        if(r > p) {
            int n = r+1;
            int x = list[r];

            int i,q = p;
            for(i=p; i<n; i++) {
                if(i<n-1 && list[i] < x) {
                    swap(list,i,q);
                    q++;
                }
                if(i==(n-1)) {
                    swap(list, i, q);
                }
            }
            partition(list,p,q,r);
        }
    }
}
